### PR TITLE
Test fwd_step_log_open

### LINKED ISSUE #2387: Test `cases_config`
Test `cases_config` using `catch2` as part of the process of replacing `util_mkdir_fopen`.

### PR DESCRIPTION
**Pull request by dafeda**
**Issue**
Resolves #2387 

Will all code under `WHEN` always run even if something crashes?
I would like to always run the following,

```C
            fwd_step_log_free(fwd_step_log);
            std::filesystem::remove_all(log_file_path);
```

but am not sure if this is the right way to go about it.

### CODE REVIEW

**Comment by eivindjahren** [comment]
> **Issue** Resolves #2387
> 
> Will all code under `WHEN` always run even if something crashes? I would like to always run the following,
> 
> ```c
>             fwd_step_log_free(fwd_step_log);
>             std::filesystem::remove_all(log_file_path);
> ```
> 
> but am not sure if this is the right way to go about it.

Unfortunately no, the way to ensure that `remove_all` is called is to do an RAII object for the log which you then use as a fixture:

```
class TmpLog
{
   string log_file_path;
   fwd_step_log_type *fwd_step_log

   public:
   TmpLog()
     : log_file_path( "tmp_fwd_step_log_open/test.txt" )
   {
       fwd_step_log = fwd_step_log_alloc();
       fwd_step_log_set_log_file(fwd_step_log, log_file_path);
       fwd_step_log_open(fwd_step_log);
   }
   ~TmpLog()
   {
       fwd_step_log_free(fwd_step_log);
       std::filesystem::remove_all(log_file_path);
   }
};
```

**Comment by dafeda** [comment]
> I believe that is the correct solution, but perhaps @dotfloat has some interesting more modern c++ thoughts on this.

Thanks!
I've pushed a version using this, but am not sure if this is following best practice.

**Comment by eivindjahren** [comment]
I certainly like it, but perhaps it would be best to ask @dotfloat for a review aswell.

**Comment by eivindjahren** [comment]
Ah, I did manage to remember something. This is still not 100% foolproof. Things can go south and the object destructor never called so the folder never removed but it would have to be really catastrophic (like pulling the cable). So I suggest you use [tmpfile](https://en.cppreference.com/w/cpp/io/c/tmpfile) so that a new temporary file is always created.

File: libres/tests/analysis/test_fwd_step_log.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#include <string>
+#include <filesystem>
+
+#include "catch2/catch.hpp"
+
+#include "ert/analysis/fwd_step_log.hpp"
+
+class TmpLog {
+public:
+    std::string log_file_path;
```
Comment (dotfloat) [inline libres/tests/analysis/test_fwd_step_log.cpp:11]:
```suggestion
    std::filesystem::path log_file_path;
```

File: libres/tests/analysis/test_fwd_step_log.cpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
+#include <iostream>
+#include <string>
+#include <filesystem>
+
+#include "catch2/catch.hpp"
+
+#include "ert/analysis/fwd_step_log.hpp"
+
+class TmpLog {
```
Comment (dotfloat) [inline libres/tests/analysis/test_fwd_step_log.cpp:9]: Wrap this class with an anonymous namespace: 
```c++
namespace {
class TmpLog {
    /* .. */
};
} // namespace
```

Anonymous namespaces are the official way to specify that a class is not to be exported out of the `.cpp` file (ie. private).

File: libres/tests/analysis/test_fwd_step_log.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+TEST_CASE_METHOD(TmpLog, "fwd_step_log_open clear_log is true", "[analysis]") {
+    bool exists_before = std::filesystem::exists(log_file_path);
+    REQUIRE_FALSE(exists_before);
+
+    fwd_step_log_set_clear_log(fwd_step_log, true);
+    fwd_step_log_open(fwd_step_log);
+
+    bool exists_after = std::filesystem::exists(log_file_path);
+    REQUIRE_FALSE(!exists_after);
```
Comment (dotfloat) [inline libres/tests/analysis/test_fwd_step_log.cpp:32]: ?
```suggestion
    REQUIRE(exists_after);
```

File: libres/tests/analysis/test_fwd_step_log.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+TEST_CASE_METHOD(TmpLog, "fwd_step_log_open clear_log is false", "[analysis]") {
+    bool exists_before = std::filesystem::exists(log_file_path);
+    REQUIRE_FALSE(exists_before);
+
+    fwd_step_log_set_clear_log(fwd_step_log, false);
+    fwd_step_log_open(fwd_step_log);
+
+    bool exists_after = std::filesystem::exists(log_file_path);
+    REQUIRE_FALSE(!exists_after);
```
Comment (dotfloat) [inline libres/tests/analysis/test_fwd_step_log.cpp:43]:
```suggestion
    REQUIRE(exists_after);
```

**Comment by dotfloat** [comment]
So, technically, you're not using RAII (at least not directly). You're using catch2's test fixture functionality. If you were to RAII, you'd write:

```c++
TEST_CASE("..", "[foo]") {
    TmpLog tmp_log

    // ...
}
```

**Comment by eivindjahren** [comment]
> So, technically, you're not using RAII (at least not directly). You're using catch2's test fixture functionality. If you were to RAII, you'd write:
> 
> ```c++
> TEST_CASE("..", "[foo]") {
>     TmpLog tmp_log;
> 
>     // ...
> }
> ```
> 
> This is probably okay. But to answer @eivindjahren , modern C++ goes full speed ahead wrt. RAII, so they are definitely the way to go.

Good to know that I am not completely out of the loop to suggest RAII here. I think we can call this RAII pattern even though we use it as a fixture. 

As you noted when we talked about this. Even calling util (which this code is fond of) will break all our efforts of trying to clean up unless we put it into a abort signal handler (lets _not_ do that). So I suggest adding that `tmpnam` is used to generate a unique name on `/tmp` everytime so that the file is never reused.

**Comment by eivindjahren** [comment]
> I agree wrt. creating a new directory. I would suggest stealing the `tmp_path_factory` code from pytest and converting it to C++.

Could be, as the original author of that I hesitated to suggest it since it might be overkill and it has a few weaknesses. Since it changes the entry point to the tests the compile time goes up a bit because catch2 is not optimized for that usage. Also we might want to resolve https://github.com/equinor/ecl/issues/821 first.

File: libres/tests/analysis/test_fwd_step_log.cpp (hunk 3 above)
Comment (dafeda) [reply libres/tests/analysis/test_fwd_step_log.cpp:32]: Boolean fun!

**Comment by dafeda** [comment]
Did not think of RAII as it's new to me.

I read up on it a bit but can't see what kind of problems we might end up having by using what we have now.
From my understanding the destructor of ´TmpLog` will get called if something fails inside `TEST_CASE_METHOD`, which is what we want.

What can go wrong?

**Comment by eivindjahren** [comment]
> Did not think of RAII as it's new to me.
> 
> I read up on it a bit but can't see what kind of problems we might end up having by using what we have now. From my understanding the destructor of ´TmpLog`will get called if something fails inside`TEST_CASE_METHOD`, which is what we want.
> 
> What can go wrong?

RAII will handle things correctly for the normal error handling stuff, however that still leaves signals and catastrophic failures. If the  program is given `SIGTERM`, `SIGABORT`, `SIGSEGV` or any of the other signals, the destuctor will not be called. If you pull out the power cord before you have a chance to perform the destructor, it will not magically happen. 

So the cases we are most worried about is `SIGABORT` which our good friend `util_abort` will do for us, `SIGSEGV` because memory violations happen all the time, and `SIGTERM` because we quite often do that via Jenkins. It is possible to install signal handlers, but that would be an incredibly flaky and complicated solution.

So the solution proposed is to instead be robust to what happens when we are unable to clean up. That involves (1) writing our temporary file to `/tmp` (or `std::filesystem::temp_directory_path` if you want to be portable, the OS manages the clean up of that) and (2) creating a unique filename every time (`std::tempnam` handles that).

pytest has a pretty cool fixture for doing this which creates a directory `/tmp/pytest-of-{$USER}/pytest-{$RUN_NUMBER}/{$TESTNAME}/` for every test run. It keeps the 5 most recent runs before it starts cleaning up. We sort-of recreated that for ecl `https://github.com/equinor/ecl/blob/master/lib/tests/tmpdir.hpp`.

**Comment by dafeda** [comment]
> Ah, I did manage to remember something. This is still not 100% foolproof. Things can go south and the object destructor never called so the folder never removed but it would have to be really catastrophic (like pulling the cable). So I suggest you use [tmpfile](https://en.cppreference.com/w/cpp/io/c/tmpfile) or [tmpnam](https://en.cppreference.com/w/cpp/io/c/tmpnam) so that a new temporary file (or directory) is always created.

Generating random names is a good idea, thanks.

clang complains about `tmpnam` though:

```console
...warning: 'tmpnam' is deprecated: This function is provided for compatibility reasons only.  Due to security concerns inherent in the design of tmpnam(3), it is highly recommended that you use mkstemp(3) instead. [-Wdeprecated-declarations]
```

`mkstemp` and `tmpfile` return handlers and what I want is a random filename.

Could perhaps create a "unique" filename by appending a timestamp or sometime similar, but I don't like that.

Ideas?

**Comment by eivindjahren** [comment]
> > Ah, I did manage to remember something. This is still not 100% foolproof. Things can go south and the object destructor never called so the folder never removed but it would have to be really catastrophic (like pulling the cable). So I suggest you use [tmpfile](https://en.cppreference.com/w/cpp/io/c/tmpfile) or [tmpnam](https://en.cppreference.com/w/cpp/io/c/tmpnam) so that a new temporary file (or directory) is always created.
> 
> Generating random names is a good idea, thanks.
> 
> clang complains about `tmpnam` though:
> 
> ```
> ...warning: 'tmpnam' is deprecated: This function is provided for compatibility reasons only.  Due to security concerns inherent in the design of tmpnam(3), it is highly recommended that you use mkstemp(3) instead. [-Wdeprecated-declarations]
> ```
> 
> `mkstemp` and `tmpfile` return handlers and what I want is a random filename.
> 
> Could perhaps create a "unique" filename by appending a timestamp or something similar, but I don't like that.
> 
> Ideas?

mkstemp has its own problem that it is posix standard only. Only thing I can think of to remedy what you are looking at is to roll your own like the fixture used in ecl.

**Comment by dafeda** [comment]
> If we include boost we could do
> 
> ```
> using namespace boost::filesystem;
> 
> path ph = temp_directory_path() / unique_path();
> create_directories(ph);
> ```

Seems a bit overkill to include `boost`, but maybe..
What if we create a filename using a uuid using for example this:

https://conan.io/center/stduuid

**Comment by eivindjahren** [comment]
> > If we include boost we could do
> > ```
> > using namespace boost::filesystem;
> > 
> > path ph = temp_directory_path() / unique_path();
> > create_directories(ph);
> > ```
> 
> Seems a bit overkill to include `boost`, but maybe.. What if we create a filename using a uuid using for example this:
> 
> https://conan.io/center/stduuid

I don't know, we might want to use boost more in any case, and if we use `stduuid` then we have to be responsible for platform specifics of whether the filename created is valid. I think that is not going to be a problem though, so I have nothing against that.

**Comment by eivindjahren** [comment]
Should be noted that the reason not to use `tmpnam` is because of a race condition (ie. two processes creating the same tempfile name because they notice it is available before either has had a chance to claim it), so we are recreating that same scenario now.

**Comment by dafeda** [comment]
This function has been deleted, but the work can be used in testing other functions that write to disk.

**Pull request closed without merging**